src/Type/Php/FilterVarDynamicReturnTypeExtension.php

```diff
@@ -74,7 +94,29 @@ public function getTypeFromFunctionCall(
 
 		$filterName = (string) $filterExpr->name;
 
+		if ($this->isNullableType($filterName, $functionCall->args[2] ?? null)) {
+			return $this->nullableTypes[$filterName];
+		}
+
 		return $this->filterTypesHashMaps[$filterName] ?? $mixedType;
 	}
 
+	private function isNullableType(string $filterName, ?Node\Arg $thirdArg): bool
+	{
+		if ($thirdArg === null || !array_key_exists($filterName, $this->nullableTypes)) {
+			return false;
+		}
+
+		$expr = $thirdArg->value;
```

Comment:
This block should be rewritten to not just use the AST, but the type system. It will work better with this kind of code:

```php
$filter = FILTER_NULL_ON_FAILURE;
filter_var($mixed, FILTER_VALIDATE_INT, $filter);
```

You can start by:

`$exprType = $scope->getType($thirdArg->value);`

And go from there. `$exprType instanceof ConstantArrayType`, go through `getValueTypes()`. Instead of `String_` node, you need `ConstantStringType`, `getValue()`.

Instead of the last line, you should check what type the constant `FILTER_NULL_ON_FAILURE` is and check that `$exprType->equals($filterNullOnFailureType)`.

Constant type can be checked by asking for `ReflectionProvider` in the constructor, asking for `$this->reflectionProvider->getConstant()` and then getting `getValueType()` from `GlobalConstantReflection`.